### ISSUE TITLE
register_full_backward_hook or register_backward_hook from pytorch

### ISSUE THREAD

**Issue by jfulem** [dotnet/TorchSharp#1059]
Please, are these module methods for registering the backward hooks implemented? Or do you plan to add those? Is it possible to add a hook to TorchNet Tensor like in Pytorch? 

I tried to use `register_forward_hook` but it looks the hook is never triggered. Please, is there any example how this can be used?

**Comment by NiklasGustafsson** [comment]
> I tried to use `register_forward_hook` but it looks the hook is never triggered. Please, is there any example how this can be used?

Do you use `forward()` or `call()` to invoke the logic of the model. The difference between the two (just like `model.forward()` vs. `model()` in PyTorch) is that only the latter invokes the hooks.

**Comment by jfulem** [comment]
OK, I used `forward()`. I will try `call()` instead. Thank you.

**Comment by NiklasGustafsson** [comment]
This is the implementation of `call()`, which isn't even virtual -- `forward()` is what you override in custom modules:

```C#
public TResult call(T input)
{
    // Call pre-hooks, if available.

    foreach (var hook in pre_hooks.Values) {
        var modified = hook(this, input);
        if (modified is not null)
            input = modified;
    }

    var result = forward(input);

    // Call post-hooks, if available.

    foreach (var hook in post_hooks.Values) {
        var modified = hook(this, input, result);
        if (modified is not null)
            result = modified;
    }

    return result;
}
```

**Comment by jfulem** [comment]
Perfect. It does work! Thank you very much. 

Please, is there any way to also register the backward hook so I can collect the gradient from a certain layer?

**Comment by NiklasGustafsson** [comment]
The backward hooks are, as far as I can tell, much more complicated to implement, and I think they have to be done in native code calling back to managed code. I haven't taken the time to figure out how to do that, yet. Do you need to manipulate the gradients during back-prop?

**Comment by jfulem** [comment]
I need them for GradCAM visualization. But it can be done without the hook I would say. But with the hook it would more convenient. Thanks.

**Comment by NiklasGustafsson** [comment]
You can always get the gradients after backprop, by using model.named_parameters() and get the `grad' property (method??) from each parameter. I think the backward hook is **necessary** only if you need to do something with it during backprop, but I may not properly understand how GradCAM works.

**Comment by jfulem** [comment]
Yep, you are correct and also that's the way I did it. Thank.
